.github/workflows/desktop.yml

```diff
@@ -0,0 +1,160 @@
+        exclude:
+        - os: ubuntu-latest
+          generator: "Visual Studio 16 2019"
+        - os: macos-latest
+          generator: "Visual Studio 16 2019"
+        - os: windows-latest
+          generator: "Unix Makefiles"
+        - os: windows-latest
+          generator: "Ninja"
+
+        include:
+        - os: windows-latest
+          architecture: "x64"
+          vcpkg_triplet: "x64-windows-static"
+          msbuild_platform: "x64"
+        - os: ubuntu-latest
+          architecture: "x64"
+          vcpkg_triplet: "x64-linux"
+        - os: macos-latest
+          architecture: "x64"
+          vcpkg_triplet: "x64-osx"
+
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          submodules: true
+
+      - name: Set env variables for subsequent steps (all)
+        run: |
+          echo "::set-env name=VCPKG_RESPONSE_FILE::external/vcpkg_${{ matrix.vcpkg_triplet }}_response_file.txt"
+          echo "::set-env name=PIP_REQUIREMENTS_FILE::external/pip_requirements.txt"
+          echo "::set-env name=MATRIX_UNIQUE_NAME::${{ matrix.os }}-${{ matrix.build_type }}-${{ matrix.architecture }}-${{ matrix.python_version }}-${{ matrix.generator }}"
+          echo "test if ccache cache works"
+
+      - name: Install Ninja (windows)
+        if: startsWith(matrix.generator, 'Ninja') && startsWith(matrix.os, 'windows')
+        run: |
+          choco install ninja
+
+      - name: Install Ninja (mac)
+        if: startsWith(matrix.generator, 'Ninja') && startsWith(matrix.os, 'macos')
+        run: |
+          brew install ninja
+
+      - name: Install Ninja (linux)
+        if: startsWith(matrix.generator, 'Ninja') && startsWith(matrix.os, 'ubuntu')
+        run: |
+          sudo apt install ninja-build
+
+      - name: Install system level applications (mac)
+        if: startsWith(matrix.os, 'macos')
+        id: macos-brew-install
+        run: |
+          brew install ccache
+
+      - name: Install system level applications (linux)
+        if: startsWith(matrix.os, 'ubuntu')
+        id: ubuntu-brew-install
+        run: |
+          sudo apt install ccache
+
+      - name: Cache vcpkg C++ dependencies
+        id: cache_vcpkg
+        uses: actions/cache@v2
+        with:
+          path: external/vcpkg/installed
+          key: dev-vcpkg-${{ matrix.vcpkg_triplet }}-${{ hashFiles(format('{0}', env.VCPKG_RESPONSE_FILE)) }}-${{ hashFiles('.git/modules/external/vcpkg/HEAD') }}
+
+      - name: Install vcpkg packages (windows)
+        if: steps.cache_vcpkg.outputs.cache-hit!='true' && startsWith(matrix.os, 'windows')
+        run: |
+          ${{ github.workspace }}/external/vcpkg/bootstrap-vcpkg.bat -disableMetrics
+          ${{ github.workspace }}/external/vcpkg/vcpkg install @${{env.VCPKG_RESPONSE_FILE}}
+
+      - name: Install vcpkg packages (mac and linux)
+        if: steps.cache_vcpkg.outputs.cache-hit!='true' && startsWith(matrix.os, 'windows')!='true'
```

Comment:
Right now we're not doing any cross compilation, but will doing a check on the OS for 'not windows' be a problem later if we add other OS targets like Android and iOS that might be running on a host Windows VM?  Or maybe there should be a host_os and target_os in the matrix?

Reply:
Excellent point. Yes, let me look into this. I am certain I can set a global environment variable during the installation but I am not sure if I can use that in an "if" expression. Going to give this a try.

Reply:
So I was planning to have a separate workflow for Android and iOS. We should discuss how we organize these github workflows. Because of similarity in steps, I was thinking of organizing them as follows,

- Desktop workflow (builds for windows/mac/linux desktop)
- Android workflow (windows/linux/mac builds for Android)
- iOS workflow (mac build for iOS)

Reply:
Ah right, this file itself is in the "desktop" classifcation.  Got it